### PR TITLE
Remove bad masquerades from cache, don't get hung up on bad cached masquerades

### PR DESCRIPTION
**Pull request by oxtoacart**
I noticed a few issues related to caching.

1. If the cache contains masquerades that end up not working, we never remove them
2. If the cache contains a lot of bad masquerades, we'll end up trying all of those before moving on to fresh ones

- [x] Do the tests pass? Consistently?

### CODE REVIEW

File: context.go [hunk 1]
```diff
@@ -100,17 +101,13 @@ func (fctx *FrontingContext) ConfigureWithHello(pool *x509.CertPool, providers m
...
-		numberToVet -= d.initCaching(cacheFile)
-	}
-
 	d.loadCandidates(d.providers)
-	if numberToVet > 0 {
-		d.vet(numberToVet)
-	} else {
-		log.Debugf("Not vetting any masquerades for %s context because we have enough cached ones", fctx.name)
-		d.signalReady()
+	d.vet(numberToVetInitially)
```
Comment (oxtoacart) [inline context.go:105]: We go ahead and vet anyway, even if there are cached masquerades. Just in case.

File: direct.go [hunk 2]
```diff
@@ -311,78 +316,88 @@ func cloneRequestWith(req *http.Request, frontedHost string, body io.ReadCloser)
...
-			case in <- m:
+			case vetted <- m:
 			default:
 				log.Debug("Dropping masquerade: retry channel full")
 			}
 		}
 	}()
 
 	for {
+		// order of preference vetted -> cached -> unvetted
```
Comment (oxtoacart) [inline direct.go:340]: So what we do now is we always prefer vetted masquerades over cached ones, but we prefer cached over unvetted masquerades.

File: direct.go [hunk 3]
```diff
@@ -311,78 +316,88 @@ func cloneRequestWith(req *http.Request, frontedHost string, body io.ReadCloser)
...
-				}
 
-				return good
-			}
 			return conn, &m, masqueradeGood, true, err
 		} else if retriable {
 			retryLater = append(retryLater, m)
 		} else {
 			log.Debugf("Dropping masquerade: non retryable error: %v", err)
+			masqueradeGood(false)
```
Comment (oxtoacart) [inline direct.go:397]: On non-retriable error, mark masquerade as bad so that we vet another.

Comment (forkner) [reply direct.go:397]: I notice we used to proactively kick off a vetOneUntilGood() in a goroutine from masqueradeGood(false). I'm not sure where this new vetting is triggered now.

Are we relying on dialWith eventually/lazily selecting unvetted items to do the vetting here?

Comment (oxtoacart) [reply direct.go:397]: Oh good catch, that needs to still happen. Fixed.

**Comment by myleshorton** [comment]
I haven't had a chance to get to this one yet, but @forkner if you're good with it, we can roll it out.

File: direct.go (hunk 3 above)
Comment (forkner) [reply direct.go:397]: might be a good candidate for a new test of some sort, though I think it would have still muddled through heh.

**Review by forkner — APPROVED** [review]
I'll try to give it a run-through at some point, but this looks fine to me.

**Comment by oxtoacart** [comment]
Thanks @forkner!

**Pull request merged**